### PR TITLE
JA: added site/ja/tutorials/text/image_captioning.ipynb

### PR DESCRIPTION
**Pull request by masa-ita**

### CODE REVIEW

File: site/ja/tutorials/text/image_captioning.ipynb [hunk 1]
```diff
@@ -0,0 +1,1137 @@
...
+    "colab_type": "text",
+    "id": "VJZXyJco6uLO"
+   },
+   "source": [
+    "# 次のステップ\n",
+    "\n",
+    "おめでとうございます！アテンション付きの画像キャプショニングモデルの訓練が終わりました。つぎは、この例 [Neural Machine Translation with Attention](../sequences/nmt_with_attention.ipynb) を御覧ください。これはおなじアーキテクチャをつかってスペイン語と英語の間の翻訳を行います。このノートブックのコードをつかって、別のデータセットで訓練を行うこともできます。"
+   ]
+  },
+  {
```
Comment (AseiSugiyama) [inline site/ja/tutorials/text/image_captioning.ipynb]: Could you remove this empty cell?

File: site/ja/tutorials/text/image_captioning.ipynb [hunk 2]
```diff
@@ -0,0 +1,1137 @@
...
+    "\n",
+    "name_of_zip = 'train2014.zip'\n",
+    "if not os.path.exists(os.path.abspath('.') + '/' + name_of_zip):\n",
+    "  image_zip = tf.keras.utils.get_file(name_of_zip,\n",
+    "                                      cache_subdir=os.path.abspath('.'),\n",
+    "                                      origin = 'http://images.cocodataset.org/zips/train2014.zip',\n",
+    "                                      extract = True)\n",
+    "  PATH = os.path.dirname(image_zip)+'/train2014/'\n",
+    "else:\n",
+    "  PATH = os.path.abspath('.')+'/train2014/'"
```
[the remark this answers is not in the document]
Comment (AseiSugiyama) [reply site/ja/tutorials/text/image_captioning.ipynb]: I created PR about this issue https://github.com/tensorflow/docs/pull/1345 .

File: site/ja/tutorials/text/image_captioning.ipynb [hunk 3]
```diff
@@ -0,0 +1,1137 @@
...
+    "\n",
+    "_[画像ソース](https://commons.wikimedia.org/wiki/Surfing#/media/File:Surfing_in_Hawaii.jpg); ライセンス: パブリックドメイン_\n",
+    "\n",
+    "これを達成するため、アテンションベースのモデルを用います。これにより、キャプションを生成する際にモデルが画像のどの部分に焦点を当てているかを見ることができます。\n",
+    "\n",
+    "![Prediction](https://tensorflow.org/images/imcap_prediction.png)\n",
+    "\n",
+    "モデルの構造は、[Show, Attend and Tell: Neural Image Caption Generation with Visual Attention](https://arxiv.org/abs/1502.03044) とおなじです。\n",
+    "\n",
+    "このノートブックは、最初から最後までが含まれた例です。このノートブックを実行すると、[MS-COCO](http://cocodataset.org/#home) データセットをダウンロードし、Inception V3 を使って画像のサブセットを前処理し、キャッシュします。その後、エンコーダー・デコーダーモデルを訓練し、訓練したモデルを使って新しい画像のキャプションを生成します。\n",
```
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/text/image_captioning.ipynb]: Thanks for reviews.
This is far better.

File: site/ja/tutorials/text/image_captioning.ipynb (hunk 2 above)
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/text/image_captioning.ipynb]: Thank you.

File: site/ja/tutorials/text/image_captioning.ipynb [hunk 4]
```diff
@@ -0,0 +1,1137 @@
...
+    "\n",
+    "![Man Surfing](https://tensorflow.org/images/surf.jpg)\n",
+    "\n",
+    "_[画像ソース](https://commons.wikimedia.org/wiki/Surfing#/media/File:Surfing_in_Hawaii.jpg); ライセンス: パブリックドメイン_\n",
+    "\n",
+    "これを達成するため、アテンションベースのモデルを用います。これにより、キャプションを生成する際にモデルが画像のどの部分に焦点を当てているかを見ることができます。\n",
+    "\n",
+    "![Prediction](https://tensorflow.org/images/imcap_prediction.png)\n",
+    "\n",
+    "モデルの構造は、[Show, Attend and Tell: Neural Image Caption Generation with Visual Attention](https://arxiv.org/abs/1502.03044) とおなじです。\n",
```
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/text/image_captioning.ipynb]: Thanks. I've changed it to "と類似のものです".

File: site/ja/tutorials/text/image_captioning.ipynb [hunk 5]
```diff
@@ -0,0 +1,1137 @@
...
+    "for annot in annotations['annotations']:\n",
+    "    caption = '<start> ' + annot['caption'] + ' <end>'\n",
+    "    image_id = annot['image_id']\n",
+    "    full_coco_image_path = PATH + 'COCO_train2014_' + '%012d.jpg' % (image_id)\n",
+    "\n",
+    "    all_img_name_vector.append(full_coco_image_path)\n",
+    "    all_captions.append(caption)\n",
+    "\n",
+    "# captions と image_names を一緒にシャッフル\n",
+    "# ランダム状態を設定\n",
```
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/text/image_captioning.ipynb]: Thank you.  I fixed.

File: site/ja/tutorials/text/image_captioning.ipynb [hunk 6]
```diff
@@ -0,0 +1,1137 @@
...
+    "\n",
+    "1. [tqdm](https://github.com/tqdm/tqdm) をインストールします:\n",
+    "\n",
+    "    `!pip install tqdm`\n",
+    "\n",
+    "2. tqdm をインポートします:\n",
+    "\n",
+    "    `from tqdm import tqdm`\n",
+    "\n",
+    "3. 下記の行を:\n",
```
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/text/image_captioning.ipynb]: Thanks.

File: site/ja/tutorials/text/image_captioning.ipynb [hunk 7]
```diff
@@ -0,0 +1,1137 @@
...
+   },
+   "outputs": [],
+   "source": [
+    "# これらのパラメータはシステム構成に合わせて自由に変更してください\n",
+    "\n",
+    "BATCH_SIZE = 64\n",
+    "BUFFER_SIZE = 1000\n",
+    "embedding_dim = 256\n",
+    "units = 512\n",
+    "vocab_size = len(tokenizer.word_index) + 1\n",
```
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/text/image_captioning.ipynb]: Thanks.  I missed this.

File: site/ja/tutorials/text/image_captioning.ipynb [hunk 8]
```diff
@@ -0,0 +1,1137 @@
...
+   },
+   "source": [
+    "## 訓練\n",
+    "\n",
+    "* それぞれの `.npy` ファイルに保存されている特徴量を取り出し、エンコーダに渡す\n",
+    "* エンコーダの出力と（0 で初期化された）隠れ状態と、デコーダの入力（開始トークン）をデコーダに渡す\n",
+    "* デコーダは予測値とデコーダの隠れ状態を返す\n",
+    "* デコーダの隠れ状態はモデルに戻され、予測値をつかって損失を計算する\n",
+    "* デコーダの次の入力を決めるために teacher forcing を用いる\n",
+    "* Teacher forcing は、デコーダの次の入力として正解単語を渡す手法である\n",
```
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/text/image_captioning.ipynb]: Thanks.  Fixed.

File: site/ja/tutorials/text/image_captioning.ipynb [hunk 9]
```diff
@@ -0,0 +1,1137 @@
...
+    "id": "sqgyz2ANKlpU"
+   },
+   "outputs": [],
+   "source": [
+    "@tf.function\n",
+    "def train_step(img_tensor, target):\n",
+    "  loss = 0\n",
+    "\n",
+    "  # バッチごとに隠れ状態を初期化\n",
+    "  # 画像が変わればキャプションも変わるため\n",
```
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/text/image_captioning.ipynb]: Thank you.  It's better.

File: site/ja/tutorials/text/image_captioning.ipynb [hunk 10]
```diff
@@ -0,0 +1,1137 @@
...
+    "    total_loss = 0\n",
+    "\n",
+    "    for (batch, (img_tensor, target)) in enumerate(dataset):\n",
+    "        batch_loss, t_loss = train_step(img_tensor, target)\n",
+    "        total_loss += t_loss\n",
+    "\n",
+    "        if batch % 100 == 0:\n",
+    "            print ('Epoch {} Batch {} Loss {:.4f}'.format(\n",
+    "              epoch + 1, batch, batch_loss.numpy() / int(target.shape[1])))\n",
+    "    # 後ほどグラフ化するためにエポック終了時の損失を保存\n",
```
[the remark this answers is not in the document]
Comment (masa-ita) [reply site/ja/tutorials/text/image_captioning.ipynb]: Thank you．It's better.

File: site/ja/tutorials/text/image_captioning.ipynb (hunk 1 above)
Comment (masa-ita) [reply site/ja/tutorials/text/image_captioning.ipynb]: Thanks.  Removed.

**Comment by masa-ita** [comment]
@AseiSugiyama 
Thank you for your review.  I've updated the notebook.  Could you please check it?

**Comment by AseiSugiyama** [comment]
@masa-ita Thank you for your great work! LGTM!

**Comment by Nabagata** [comment]
@masa-ita @lamberta @AseiSugiyama  I'm trying to include the metrics like BLEU for inference purpose during training. Can the metric implementation be added in the original notebook?

**Comment by lamberta** [comment]
Hi @Nabagata, the source notebook is here: [site/en/tutorials/text/image_captioning.ipynb](https://github.com/tensorflow/docs/blob/master/site/en/tutorials/text/image_captioning.ipynb), you can send a PR there.

I believe @yashk2810 was the original author and might have an opinion about including the metric implementation

**Comment by yashk2810** [comment]
BLEU inference sounds good to me. You may wanna add that to nmt_with_attention first though since that notebook is fast and easy to see the improvements.

**Comment by Nabagata** [comment]
Hi @yashk2810, I got this error when I used the same notebook on my system

`W tensorflow/core/grappler/optimizers/implementation_selector.cc:310] Skipping optimization due to error while loading function libraries: Invalid argument: Functions '__inference___backward_cudnn_gru_with_fallback_177892_178030_specialized_for_StatefulPartitionedCall_1_at___inference_train_step_211166' and '__inference___backward_standard_gru_178852_179228' both implement 'gru_5657f23c-9806-412f-8be8-fdd4216af8b9' but their signatures do not match.
`
This led to epochs getting executed way faster than it took on Colab.
Also, I'd like to know how does the evaluate function used the stored checkpoints from the folder. Didn't understand how it's happening.

**Comment by lamberta** [comment]
@Nabagata Since this pull requests has been /merged/closed, can you please create a PR on the tensorflow/docs repo, or a docs issue in the tensorflow/tensorflow repo? Thanks
